src/datafusion/table_provider.rs

```diff
@@ -31,55 +32,163 @@ use datafusion::{
-        let schema = zarr_path.get_zarr_metadata().await?.arrow_schema()?;
+        let n_partitions = self.table_partition_cols.len();
+        let mut files = table_path.list_all_files(state, &store, "zgroup").await?;
+        let mut schema_to_return: Option<Schema> = None;
+        while let Some(file) = files.next().await {
+            let mut p = prefix.clone();
+            let file = file?.location;
+            for (cnt, part) in file.prefix_match(prefix).unwrap().enumerate() {
+                if cnt == n_partitions {
+                    if let Some(ext) = file.extension() {
+                        if ext == "zgroup" {
+                            let schema = ZarrPath::new(store.clone(), p.clone())
+                                .get_zarr_metadata()
+                                .await?
+                                .arrow_schema()?;
+                            if let Some(sch) = &schema_to_return {
+                                if sch != &schema {
+                                    return Err(ZarrError::InvalidMetadata(
+                                        "mismatch between different partition schemas".into(),
+                                    ));
+                                }
+                            } else {
+                                schema_to_return = Some(schema);
+                            }
+                        }
+                    }
+                }
+                p = p.child(part);
+            }
+        }
 
-        Ok(schema)
+        if let Some(schema_to_return) = schema_to_return {
+            return Ok(schema_to_return);
+        }
+        Err(ZarrError::InvalidMetadata(
+            "could not infer schema for zarr table path".into(),
+        ))
     }
 }
 
 pub struct ListingZarrTableConfig {
-    /// The inner listing table configuration
     table_path: ListingTableUrl,
+    pub file_schema: Schema,
+    pub options: Option<ListingZarrTableOptions>,
 }
 
 impl ListingZarrTableConfig {
     /// Create a new ListingZarrTableConfig
-    pub fn new(table_path: ListingTableUrl) -> Self {
-        Self { table_path }
+    pub fn new(
+        table_path: ListingTableUrl,
+        file_schema: Schema,
+        options: Option<ListingZarrTableOptions>,
+    ) -> Self {
+        Self {
+            table_path,
+            file_schema,
+            options,
+        }
     }
 }
 
 pub struct ZarrTableProvider {
+    // the distinction between the file schema and the table schema is
+    // that the latter could include partitioned columns.
+    file_schema: Schema,
     table_schema: Schema,
-    config: ListingZarrTableConfig,
+    table_path: ListingTableUrl,
+    options: ListingZarrTableOptions,
 }
 
 impl ZarrTableProvider {
-    pub fn new(config: ListingZarrTableConfig, table_schema: Schema) -> Self {
-        Self {
+    pub fn try_new(config: ListingZarrTableConfig) -> DataFusionResult<Self> {
+        // TODO does options need to be an Option?
```

Comment:
This is my fault, and I think you're probably right here. Somewhat recently I refactored code I had that followed this pattern to be simpler and not make options an Option.

Reply:
Right okay, I think I made that comment because I didn't see a use case for setting it to None. I thought maybe there was a reason related to other datafusion features I didn't know about, but if it's not necessary I'll just remove it.